RecoLocalCalo/EcalRecProducers/plugins/EcalCPURecHitProducer.cc

```diff
@@ -0,0 +1,158 @@
+
+// algorithm specific
+
+#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
+
+#include "CUDADataFormats/EcalRecHitSoA/interface/EcalRecHit.h"
+
+class EcalCPURecHitProducer : public edm::stream::EDProducer<edm::ExternalWork> {
+public:
+  explicit EcalCPURecHitProducer(edm::ParameterSet const& ps);
+  ~EcalCPURecHitProducer() override = default;
+  static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+private:
+  void acquire(edm::Event const&, edm::EventSetup const&, edm::WaitingTaskWithArenaHolder) override;
+  void produce(edm::Event&, edm::EventSetup const&) override;
+
+private:
+  using InputProduct = cms::cuda::Product<ecal::RecHit<calo::common::DevStoragePolicy>>;
+  edm::EDGetTokenT<InputProduct> recHitsInEBToken_, recHitsInEEToken_;
+  using OutputProduct = ecal::RecHit<calo::common::VecStoragePolicy<calo::common::CUDAHostAllocatorAlias>>;
+  edm::EDPutTokenT<OutputProduct> recHitsOutEBToken_, recHitsOutEEToken_;
+
+  OutputProduct recHitsEB_, recHitsEE_;
+  bool containsTimingInformation_;
+};
+
+void EcalCPURecHitProducer::fillDescriptions(edm::ConfigurationDescriptions& confDesc) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("recHitsInLabelEB", edm::InputTag{"ecalRecHitProducerGPU", "EcalRecHitsEB"});
+  desc.add<edm::InputTag>("recHitsInLabelEE", edm::InputTag{"ecalRecHitProducerGPU", "EcalRecHitsEE"});
+  desc.add<std::string>("recHitsOutLabelEB", "EcalRecHitsEB");
+  desc.add<std::string>("recHitsOutLabelEE", "EcalRecHitsEE");
+  desc.add<bool>("containsTimingInformation", false);
+
+  confDesc.addWithDefaultLabel(desc);
+}
+
+EcalCPURecHitProducer::EcalCPURecHitProducer(const edm::ParameterSet& ps)
+    : recHitsInEBToken_{consumes<InputProduct>(ps.getParameter<edm::InputTag>("recHitsInLabelEB"))},
+      recHitsInEEToken_{consumes<InputProduct>(ps.getParameter<edm::InputTag>("recHitsInLabelEE"))},
+      recHitsOutEBToken_{produces<OutputProduct>(ps.getParameter<std::string>("recHitsOutLabelEB"))},
+      recHitsOutEEToken_{produces<OutputProduct>(ps.getParameter<std::string>("recHitsOutLabelEE"))},
+      containsTimingInformation_{ps.getParameter<bool>("containsTimingInformation")} {}
+
+void EcalCPURecHitProducer::acquire(edm::Event const& event,
+                                    edm::EventSetup const& setup,
+                                    edm::WaitingTaskWithArenaHolder taskHolder) {
+  // retrieve data/ctx
+  auto const& ebRecHitsProduct = event.get(recHitsInEBToken_);
+  auto const& eeRecHitsProduct = event.get(recHitsInEEToken_);
+  cms::cuda::ScopedContextAcquire ctx{ebRecHitsProduct, std::move(taskHolder)};
+  auto const& ebRecHits = ctx.get(ebRecHitsProduct);
+  auto const& eeRecHits = ctx.get(eeRecHitsProduct);
+
+  // resize the output buffers
+  recHitsEB_.resize(ebRecHits.size);
+  recHitsEE_.resize(eeRecHits.size);
+
+  //     std::cout << " [EcalCPURecHitProducer::acquire] ebRecHits.size = " << ebRecHits.size << std::endl;
```

Comment:
wrap in `ifdef` per the above discussion on a common approach for GPU debugging -> commented out print functions are not preferred.